ssl/quic/quic_rx_depack.c

```diff
@@ -0,0 +1,612 @@
+            /* FALLTHRU */
+        case OSSL_QUIC_FRAME_TYPE_PADDING:
+            /*
+             * We ignore this.
+             * For the PING frame, eliciting an ACK is enough.
+             * Since the PING frame is only a header, we simply
+             * skip it.
+             */
+            ossl_quic_wire_skip_frame_header(pkt, &frame_type);
+            break;
+
+        case OSSL_QUIC_FRAME_TYPE_ACK_WITHOUT_ECN:
+        case OSSL_QUIC_FRAME_TYPE_ACK_WITH_ECN: {
+            OSSL_QUIC_FRAME_ACK ack;
+            OSSL_QUIC_ACK_RANGE *ack_ranges;
+            uint64_t total_ranges = 0;
+            int ok = 1;          /* Assume the best */
+
+            if (!ossl_quic_wire_peek_frame_ack_num_ranges(pkt, &total_ranges)
+                || (ack_ranges = OPENSSL_zalloc(sizeof(ack_ranges[0])
+                                                * total_ranges)))
+                return 0;
+
+            ack.ack_ranges = ack_ranges;
+            ack.num_ack_ranges = total_ranges;
+
+            /* QUESTION: where do I get the ack delay exponent? */
+            if (!ossl_quic_wire_decode_frame_ack(pkt,
+                                                 1 /* ACK delay exponent */,
+                                                 &ack, NULL))
+                ok = 0;
+            if (ok
+                && mode == process_frames
+                && !ossl_ackm_on_rx_ack_frame(GET_CONN_ACKM(connection), &ack,
+                                              packet_space, received))
+                ok = 0;
+
+            OPENSSL_free(ack_ranges);
+            if (!ok)
+                return 0;
+            break;
+        }
+
+        case OSSL_QUIC_FRAME_TYPE_RESET_STREAM: {
+            OSSL_QUIC_FRAME_RESET_STREAM frame_data;
+
+            if (!ossl_quic_wire_decode_frame_reset_stream(pkt, &frame_data))
+                return 0;
+
+            switch (mode) {
+            case collect_ack_data:
+                /* This frame makes the packet ACK eliciting */
+                ackm_data->is_ack_eliciting = 1;
+                break;
+            case process_frames: {
+                QUIC_STREAM *stream
+                    = ssl_get_stream(connection, frame_data.stream_id);
+                int stream_type = ssl_get_stream_type(stream);
+
+                ssl_close_stream(stream);
+                if ((stream_type & SSL_STREAM_TYPE_W) != 0) {
+                    OSSL_STATM *statm = GET_CONN_STATM(connection);
+
+                    ssl_close_connection(connection);
+                    ossl_statm_set_error_state(statm, QUIC_STREAM_STATE_ERROR);
+
+                    /*
+                     * There's no point to continue with this packet.  However,
+                     * processing it was a success, so we return with a happy
+                     * face, and leave it to the state machine what to do with
+                     * the general STREAM_STATE_ERROR state.
+                     */
+                    return 1;
```

Comment:
Hmmm. This just seems like a fatal error to me. I'm not sure why we would return 1 here. We've given up processing the packet halfway through. Just return 0 and immediately signal a failure of the whole connection, propagating this error up to the return value of `ossl_quic_depacketize`.

Reply:
Ok...  trouble is that in the `collect_ack_data` has no idea that this will fail, and if you look here, you'll see that ACK is emitted before the frames are actually processed (`process_frames` run):

https://github.com/openssl/openssl/pull/18838/files#diff-38c7635c72a873d4d53258267676d04ecf7b625177964441ec808e8f3765ce7eR602-R617

Unless I've misunderstood, it must really happen this way, i.e. ACK must be emitted before the frames are actually processed and used.

So as far as I understand, we've a bit of a catch-22, unless you're telling me that the ACK doesn't have to be emited before the frames have been actually used...  that would actually make things easier, as I wouldn't have to process the OSSL_QRX_PKT twice.

Reply:
Ok, I imagined it wasn't quite all right to do it like that...  but if that's the case, then fine, I'm game.

Reply:
Changed